go-client/client/client.go

```diff
@@ -0,0 +1,312 @@
+// This package allows you to interface with a Deephaven server over a network connection using Go.
+// It can upload, manipulate, and download tables, among other features.
+// First, use client.NewClient to connect to the server, then the Client can be used to perform operations.
+// See the provided examples in the examples/ folder or the individual code documentation for more.
+// All methods are thread-safe unless otherwise specified.
+package client
+
+import (
+	"context"
+	"errors"
+	"log"
+	"sync"
+	"sync/atomic"
+
+	applicationpb2 "github.com/deephaven/deephaven-core/go-client/internal/proto/application"
+	consolepb2 "github.com/deephaven/deephaven-core/go-client/internal/proto/console"
+	ticketpb2 "github.com/deephaven/deephaven-core/go-client/internal/proto/ticket"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/grpc/metadata"
+)
+
+// Returned as an error when trying to perform a network operation on a client that has been closed.
+var ErrClosedClient = errors.New("client is closed")
+
+// A fieldId is a unique identifier for a field on the server,
+// where a "field" could be e.g. a table or a plot.
+// Currently only tables are kept track of.
+type fieldId struct {
+	appId     string // appId is the application scope for the field. For the global scope this is "scope".
+	fieldName string // fieldName is the name of the field.
+}
+
+// Maintains a connection to a Deephaven server.
+// It can be used to run scripts, create new tables, execute queries, etc.
+// Check the various methods of Client to learn more.
+type Client struct {
+	// Guards client-wide state. This means specifically:
+	// - Is a FetchTables request running?
+	// - Is the client closed?
+	// Other functionality does not need a lock, since the gRPC interface is already thread-safe,
+	// the tables array has its own lock, and the session token also has its own lock.
+	lock sync.Mutex
+
+	grpcChannel *grpc.ClientConn
+
+	sessionStub
+	consoleStub
+	flightStub
+	tableStub
+	appStub
+	inputTableStub
+
+	// A simple counter that increments every time a new ticket is needed.
+	// Must be accessed atomically.
+	nextTicket int32
+
+	tablesLock sync.Mutex               // Guards the tables map.
+	tables     map[fieldId]*TableHandle // A map of tables that can be opened using OpenTable
+}
+
+// NewClient starts a connection to a Deephaven server.
+//
+// scriptLanguage can be either "python" or "groovy", and must match the language used on the server. Python is the default.
+//
+// The client should be closed using Close() after it is done being used.
+//
+// Note that the provided context is saved and used to send keepalive messages.
+// Keepalive messages are sent automatically by the client at a regular interval (~30 seconds) so that the connection can continue.
```

Comment:
```suggestion
// Keepalive messages are sent automatically by the client to the server at a regular interval (~30 seconds)
// so that the connection remains open.  The provided context is saved and used to send keepalive messages.
```